pkg/cpu/cpu_linux.go

```diff
@@ -37,117 +37,109 @@ func (i *Info) load() error {
 			// can we drop /proc/cpuinfo.
-			if len(procAttrs[lcoreID]["flags"]) != 0 { // x86
-				proc.Capabilities = strings.Split(procAttrs[lcoreID]["flags"], " ")
-			} else if len(procAttrs[lcoreID]["Features"]) != 0 { // ARM64
-				proc.Capabilities = strings.Split(procAttrs[lcoreID]["Features"], " ")
+			if len(lp.Attrs["flags"]) != 0 { // x86
+				proc.Capabilities = strings.Split(lp.Attrs["flags"], " ")
+			} else if len(lp.Attrs["Features"]) != 0 { // ARM64
+				proc.Capabilities = strings.Split(lp.Attrs["Features"], " ")
 			}
-			if len(procAttrs[lcoreID]["model name"]) != 0 {
-				proc.Model = procAttrs[lcoreID]["model name"]
-			} else if len(procAttrs[lcoreID]["uarch"]) != 0 { // SiFive
-				proc.Model = procAttrs[lcoreID]["uarch"]
+			if len(lp.Attrs["model name"]) != 0 {
+				proc.Model = lp.Attrs["model name"]
+			} else if len(lp.Attrs["uarch"]) != 0 { // SiFive
+				proc.Model = lp.Attrs["uarch"]
 			}
-			if len(procAttrs[lcoreID]["vendor_id"]) != 0 {
-				proc.Vendor = procAttrs[lcoreID]["vendor_id"]
-			} else if len(procAttrs[lcoreID]["isa"]) != 0 { // RISCV64
-				proc.Vendor = procAttrs[lcoreID]["isa"]
+			if len(lp.Attrs["vendor_id"]) != 0 {
+				proc.Vendor = lp.Attrs["vendor_id"]
+			} else if len(lp.Attrs["isa"]) != 0 { // RISCV64
+				proc.Vendor = lp.Attrs["isa"]
 			}
-			procs = append(procs, proc)
+			procs[procID] = proc
 		}
 
-		// Fetch Core ID
-		coreIdPath := filepath.Join(paths.SysDevicesSystemCPU, fmt.Sprintf("cpu%d", lcoreID), "topology", "core_id")
-		coreID := util.SafeIntFromFile(ctx, coreIdPath)
-		core := CoreByID(proc.Cores, coreID)
+		coreID := coreIDFromLogicalProcessorID(ctx, lpID)
+		core := proc.CoreByID(coreID)
 		if core == nil {
-			core = &ProcessorCore{Index: coreID, NumThreads: 1}
+			core = &ProcessorCore{ID: coreID, NumThreads: 1}
 			proc.Cores = append(proc.Cores, core)
 			proc.NumCores += 1
 		} else {
 			core.NumThreads += 1
 		}
 		proc.NumThreads += 1
-		core.LogicalProcessors = append(core.LogicalProcessors, lcoreID)
+		core.LogicalProcessors = append(core.LogicalProcessors, lpID)
 	}
-	return procs
+	res := []*Processor{}
+	for _, p := range procs {
+		res = append(res, p)
+	}
+	return res
+}
+
+// processorIDFromLogicalProcessorID returns the processor physical package ID
+// for the supplied logical processor ID
+func processorIDFromLogicalProcessorID(ctx *context.Context, lpID int) int {
+	paths := linuxpath.New(ctx)
+	// Fetch CPU ID
+	path := filepath.Join(
+		paths.SysDevicesSystemCPU,
+		fmt.Sprintf("cpu%d", lpID),
+		"topology", "physical_package_id",
+	)
+	return util.SafeIntFromFile(ctx, path)
+}
+
+// coreIDFromLogicalProcessorID returns the core ID for the supplied logical
+// processor ID
+func coreIDFromLogicalProcessorID(ctx *context.Context, lpID int) int {
+	paths := linuxpath.New(ctx)
+	// Fetch CPU ID
+	path := filepath.Join(
+		paths.SysDevicesSystemCPU,
+		fmt.Sprintf("cpu%d", lpID),
+		"topology", "core_id",
+	)
+	return util.SafeIntFromFile(ctx, path)
```

Comment:
unrelated to this PR: I think eventually we should evaluate if better to cache data from sysfs instead of hitting the FS layer all the time. Granted, sysfs is a pseudofs, but still